services/src/main/java/org/fao/geonet/api/records/attachments/AttachmentsApi.java

```diff
@@ -277,38 +278,60 @@ public ResponseEntity<Resource> getResource(
-            // Check if the request is not modified based on ETag or Last-Modified
-            if (new ServletWebRequest(request).checkNotModified(fileETag, fileLastModified)) {
-                return ResponseEntity.status(HttpStatus.NOT_MODIFIED).headers(responseHeaders).build();
-            }
+        // Get the resource holder for the requested resource
+        Store.ResourceHolder file = store.getResource(context, metadataUuid, resourceId, approved);
 
-            // Check if the file is an image and if a size is requested
-            if (fileMediaType.getType().equals("image") && size != null) {
-                return serveResizedImage(file.getPath(), file.getMetadata().getFilename(), responseHeaders, size);
-            }
+        // Get the resource properties
+        long fileSize = Files.size(file.getPath());
+        long fileLastModified = file.getMetadata().getLastModification().getTime();
+        MediaType fileMediaType = MediaType.parseMediaType(getFileContentType(file.getPath()));
+        String fileETag = "\"" + DigestUtils.md5Hex(file.getMetadata().getFilename() + fileLastModified + fileSize) + "\"";
 
-            // Set headers for downloads
-            responseHeaders.setContentType(fileMediaType);
-            responseHeaders.setContentDisposition(ContentDisposition.attachment().filename(file.getMetadata().getFilename()).build());
-            responseHeaders.setContentLength(fileSize);
+        // Set common headers
+        HttpHeaders responseHeaders = new HttpHeaders();
+        responseHeaders.setCacheControl(CacheControl.noCache());
+        responseHeaders.setLastModified(fileLastModified);
+        responseHeaders.setETag(fileETag);
+
+        // Check if the request is not modified based on ETag or Last-Modified
+        if (new ServletWebRequest(request).checkNotModified(fileETag, fileLastModified)) {
+            return ResponseEntity.status(HttpStatus.NOT_MODIFIED).headers(responseHeaders).build();
+        }
 
-            // Spring automatically handles range requests when returning ResponseEntity<Resource> to support resumable downloads.
-            return ResponseEntity.ok()
-                .headers(responseHeaders)
-                .body(new PathResource(file.getPath()));
+        // Check if the file is an image and if a size is requested
+        if (fileMediaType.getType().equals("image") && size != null) {
+            return serveResizedImage(file.getPath(), file.getMetadata().getFilename(), responseHeaders, size);
         }
+
+        // Set headers for downloads
+        responseHeaders.setContentType(fileMediaType);
+        responseHeaders.setContentDisposition(ContentDisposition.attachment().filename(file.getMetadata().getFilename()).build());
+        responseHeaders.setContentLength(fileSize);
+
+        // Wrap a PathResource so that closing its InputStream also closes the holder
+        // Auto-closing will not work here as it will prematurely delete the temporary file for jcloud
```

Comment:
```suggestion
        // Auto-closing will not work here as it will prematurely delete the temporary file for storage implementations that use temporary files.
```